### PR TITLE
Extended definitions contribute their variables

### PR DESCRIPTION
**Pull request by dahaic**
#### Description:

- To identify which variables are needed within particular xccdf rule, we have to delve into depths of extended definitions as well.

#### Rationale:

- PR #2748 missed one detail - extended definitions.

### CODE REVIEW

File: shared/modules/parse_oval.py [hunk 1]
```diff
@@ -25,16 +25,18 @@ class ElementFinder(object):
     def __init__(self, oval_groups):
         self.oval_groups = oval_groups
         self.target = None
+        self.attrib = None
         self.result = set()
 
-    def find_element(self, start_element, target_element_name):
+    def find_element(self, start_element, target_element_name, seeked_attrib):
```
Comment (mpreisler) [inline shared/modules/parse_oval.py]: isn't this supposed to be "sought" or something like that? I dunno but seeked looks weird.

File: shared/utils/relabel-ids.py [hunk 2]
```diff
@@ -201,15 +202,36 @@ def _add_cce_id_refs_to_oval_checks(self, idmappingdict):
                 msg = "Failed to add CCE ID to {0}.".format(ovalid)
                 raise ssgcommon.SSGError(msg)
 
+    def get_nested_definitions(self, oval_def_id):
+        processed_def_ids = set([])
```
Comment (mpreisler) [inline shared/utils/relabel-ids.py]: isn't just `set()` the same thing?

File: shared/utils/relabel-ids.py [hunk 3]
```diff
@@ -2,6 +2,7 @@
 
 from __future__ import print_function
 
+import copy
```
Comment (mpreisler) [inline shared/utils/relabel-ids.py]: this doesn't seem to be used

File: shared/modules/parse_oval.py (hunk 1 above)
Comment (dahaic) [reply shared/modules/parse_oval.py]: Yeah,  you are probably right. I do not connect technical "seek" with the proper English word. Fixed ;)

File: shared/utils/relabel-ids.py (hunk 2 above)
Comment (dahaic) [reply shared/utils/relabel-ids.py]: Huh, it never occurred to me :D

File: shared/utils/relabel-ids.py (hunk 3 above)
Comment (dahaic) [reply shared/utils/relabel-ids.py]: Good catch, fixed :)

File: shared/utils/relabel-ids.py [hunk 4]
```diff
@@ -201,15 +202,36 @@ def _add_cce_id_refs_to_oval_checks(self, idmappingdict):
...
+                definition = queue.pop()
+            except KeyError:
+                break
+            processed_def_ids.add(definition)
+            definition_tree = self.oval_groups["definitions"].get(definition)
+            extensions = parse_oval.find_extending_defs(self.oval_groups,
+                                                        definition_tree)
+            if not extensions:
+                continue
+            queue |= extensions - processed_def_ids
```
[the remark this answers is not in the document]
Comment (dahaic) [reply shared/utils/relabel-ids.py]:
```
queue = queue.union(extensions.difference(processed_def_ids))
```
I would rather not use this ^^, it takes me much longer to realize what's going on.

File: shared/utils/relabel-ids.py [hunk 5]
```diff
@@ -201,15 +201,36 @@ def _add_cce_id_refs_to_oval_checks(self, idmappingdict):
                 msg = "Failed to add CCE ID to {0}.".format(ovalid)
                 raise ssgcommon.SSGError(msg)
 
+    def get_nested_definitions(self, oval_def_id):
+        processed_def_ids = set()
+        queue = set([oval_def_id])
+        while True:
```
Comment (matejak) [inline shared/utils/relabel-ids.py]: You can have `while queue:` here, and get rid of the try/except when popping.

File: shared/utils/relabel-ids.py [hunk 6]
```diff
@@ -201,15 +201,36 @@ def _add_cce_id_refs_to_oval_checks(self, idmappingdict):
...
+            try:
+                definition = queue.pop()
```
Comment (matejak) [inline shared/utils/relabel-ids.py]: I would call that `definition_id`, or `def_id`, as you have the former on other places.

File: shared/utils/relabel-ids.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (matejak) [reply shared/utils/relabel-ids.py]: A compromise may be `queue.update(extensions - processed_def_ids)`.

File: shared/utils/relabel-ids.py (hunk 5 above)
Comment (dahaic) [reply shared/utils/relabel-ids.py]: Thanks, looks much better.

File: shared/utils/relabel-ids.py (hunk 6 above)
Comment (dahaic) [reply shared/utils/relabel-ids.py]: Good point

File: shared/utils/relabel-ids.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (dahaic) [reply shared/utils/relabel-ids.py]: It's just that with my line, I know immediately it's `set` operation. `update` is a bit vague. If you both think that either of your ways is more readable for general public, I will change it. But it's going to be a downgrade for me personally.

**Comment by matejak** [comment]
Great contribution!

**Pull request merged**